### DISCUSSION TITLE
Need to truncate a segment

### DISCUSSION THREAD

**Discussion by kgarland9363** [nextgenhealthcare/connect#6270] [Q&A] [answered]
I am trying to truncate the AL1.5.1 segment to only allow 250 characters to be sent. Once it's truncated I need to add some text to direct users to the patient's record for the complete value. I have tried everything I can think of to set this up but nothing I set up is reflected in my transformed message. Does anyone know I can accomplish this. Below is a test message showing >250 characters in the AL1.5.1 segment. I am also including the Javascript code I have so far.

**Copy of AL1 segment (this is test data):** 
AL1|1|DRUG INGREDI|3888^CODEINE^NONFOOD-ALLERGY||Sdfjsdokfpodk fsopdk fsopadk fpofspfkopsdeopefasopef posfkpofisfjjjffjslkdfmdlkdklk opcfkdrwecrk wsdkfselksfj sd;kfmewklfnklwefeflfasesefjk sdkldklklsdmfvklsdmf;lsaf;skf;lsdkfsl;mkfl;sdkl;dk;flsdkfskfeawer;I fpsdkvl;dsdopfjkdsorwe;lrfl;kds xcvkxl;vmdf;fsdk sdklfl;rfefjkdkvdf;lamkp;kfdkgdmklferfg['erprtfeore0434optkpdfkgl;dfl;fkdp mikopertvpoe4imgvev4g45tvpowe4roprerfkfsdl;fsdkl;sdfl;ldkffkdkfl;sdfkseopsertekoroprpeorrfsdpfpsdfkl;dksfl;klgkopt0o4rtk kgdfk p;vbkkpogkdogopdkp[sdktfgdiopvboprektersgpfpgklje|20240715
AL1|2|Drug Class|25^PENICILLINS^NONFOOD-ALLERGY||Sddsffdssfsdfsfsdfsdfsdfsdfsdfsdfdfdsfsdfdfsfdsfsfsdfsdfsdfsdfdfsdfsdfsdfsfdfsfsdasfdfsdfsdfsdfdsdssdfdfsfd Do you see this|20240715

**Copy of current code:**
// Function to truncate and append message if necessary
function truncateAndAppend(segment, maxLength, appendMessage) {
    if (segment.length > maxLength) {
        return segment.substring(0, maxLength) + " " + appendMessage;
    } else {
        return segment;
    }
}

// Initialize variables
var maxLength = 250;
var appendMessage = "See EPIC record for additional allergy information";

// Access the raw message
var rawMessage = message;

// Extract the AL1 segments
var al1Segments = rawMessage.split('\n').filter(function(line) {
    return line.startsWith("AL1");
});

// Process each AL1 segment
var newAl1Segments = [];
for (var i = 0; i < al1Segments.length; i++) {
    var al1Segment = al1Segments[i];
    var fields = al1Segment.split('|');
    if (fields.length > 5) {
        var originalText = fields[5];
        var truncatedText = truncateAndAppend(originalText, maxLength, appendMessage);
        fields[5] = truncatedText;
        newAl1Segments.push(fields.join('|'));
    } else {
        newAl1Segments.push(al1Segment);
    }
}

// Replace old AL1 segments with new ones
var updatedMessage = rawMessage;
for (var i = 0; i < al1Segments.length; i++) {
    updatedMessage = updatedMessage.replace(al1Segments[i], newAl1Segments[i]);
}

// Update the raw message with the modified content
message = updatedMessage;

**Comment by pacmano1** [comment]
Please use code formatting and highlighting here.  Please read https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks. 

Fix your post please.

**Reply by pacmano1** [reply, marked as answer]
This can be shortened up of course.  
```JS
var maxLength = 250;
var appendText = "test text some more stuff here";

for each (var allergy in msg.AL1) {
	var this_allergy_desc = allergy['AL1.5']['AL1.5.1'].toString()
	if (this_allergy_desc.length > maxLength) {
	    allergy['AL1.5']['AL1.5.1'] = this_allergy_desc.substring(0, maxLength) + appendText;
	}
}
```

**Reply by rjkroll** [reply]
I was able to get it to work via the following.   If the segment is > 250 characters, it will truncate and append the mesage to see EPIC such that the total length, with the appended message, should still be less than 250 charters.    

```
const maxLength = 250;
const appendMessage = "See EPIC record for additional allergy information";

function truncateAndAppend(text, maxLength, appendMessage) {
  const totalLength = maxLength - appendMessage.length - 1; 
  if (text.length > totalLength) {
    return text.slice(0, totalLength) + " " + appendMessage;
  }
  return text;
}

var al1Segments = msg['AL1'];

for (var i = 0; i < al1Segments.length(); i++) {
  var al1Segment = al1Segments[i];
  
  var allergyText = al1Segment['AL1.5']['AL1.5.1'].toString();
  al1Segment['AL1.5']['AL1.5.1'] = truncateAndAppend(allergyText, maxLength, appendMessage);
}

return msg;
```

Which outputted the following:

```
AL1|1|DRUG INGREDI|3888^CODEINE^NONFOOD-ALLERGY||Sdfjsdokfpodk fsopdk fsopadk fpofspfkopsdeopefasopef posfkpofisfjjjffjslkdfmdlkdklk opcfkdrwecrk wsdkfselksfj sd;kfmewklfnklwefeflfasesefjk sdkldklklsdmfvklsdmf;lsaf;skf;lsdkfsl;mkfl;sdkl;dk;flsdkfsk See EPIC record for additional allergy information|20240715
AL1|2|Drug Class|25^PENICILLINS^NONFOOD-ALLERGY||Sddsffdssfsdfsfsdfsdfsdfsdfsdfsdfdfdsfsdfdfsfdsfsfsdfsdfsdfsdfdfsdfsdfsdfsfdfsfsdasfdfsdfsdfsdfdsdssdfdfsfd Do you see this|20240715
```

**Comment by kgarland9363** [comment]
Thank you, thank you everyone for your input. I have played with your suggestions and was able to get it to work properly. I sincerely appreciate your input.

**Reply by pacmano1** [reply]
Please fix your original post.
